go/mysql/conn.go

```diff
@@ -1210,19 +1239,108 @@ func (c *Conn) handleNextCommand(handler Handler) error {
+				log.Error("Error writing error packet to client: %v", err)
+				return err
+			}
+		}
+
+		c.startWriterBuffering()
+
+		stmtID, numRows, ok := c.parseComStmtFetch(data)
 		c.recycleReadPacket()
+		if !ok {
+			log.Error("Got unhandled packet from client %v, returning error: %v", c.ConnectionID, data)
+			if err := c.writeErrorPacket(ERUnknownComError, SSUnknownComError, "error handling packet: %v", data); err != nil {
+				log.Error("Error writing error packet to client: %v", err)
+				return err
+			}
+		}
 
-		err = fmt.Errorf("COM_STMT_FETCH is not suppported")
-		if werr := c.writeErrorPacketFromError(err); werr != nil {
-			// If we can't even write the error, we're done.
-			log.Errorf("Error writing query error to %s: %v", c, werr)
-			return werr
+		if stmtID != c.cs.stmtID {
+			log.Errorf("Requested stmtID does not match stmtID of open cursor. Client %v, returning error: %v", c.ConnectionID, data)
+			if err := c.writeErrorPacket(ERUnknownComError, SSUnknownComError, "error handling packet: %v", data); err != nil {
+				log.Error("Error writing error packet to client: %v", err)
+				return err
+			}
+		}
+
+		// If fetching, we definitely have a cursor open
+		c.StatusFlags |= uint16(ServerCursorExists)
+
+		// while we are still expecting rows
+		for !c.cs.noData && (c.cs.pending == nil || len(c.cs.pending.Rows) < int(numRows)) {
+			// drain rows from next to cs.pending
+			var newqr *sqltypes.Result
+			newqr, ok = <-c.cs.next
+
+			// channel is closed, meaning there are no more rows
+			if !ok {
+				// check if there was an error
+				if err = <- c.cs.done; err != nil {
+					return err
+				}
+				// indicate that there is no more data, and prevent blocking from reading c.cs.done for errors again
+				c.cs.noData = true
+				break
+			}
+
+			// move rows into buffer
+			if c.cs.pending == nil {
+				c.cs.pending = newqr
+			} else {
+				c.cs.pending.Rows = append(c.cs.pending.Rows, newqr.Rows...)
+			}
+
+			// write out some rows to prevent packets from getting too large
+			for len(c.cs.pending.Rows) < int(numRows) && len(c.cs.pending.Rows) > 128 {
+				pendingRows := c.cs.pending.Rows[128:]
+				c.cs.pending.Rows = c.cs.pending.Rows[:128]
+				if err = c.writeBinaryRows(c.cs.pending); err != nil {
+					log.Errorf("Error writing result to %s: %v", c, err)
+					return err
+				}
+				c.cs.pending.Rows = pendingRows
+				numRows -= 128
+			}
+		}
+
+		// cap the number of rows
+		if int(numRows) > len(c.cs.pending.Rows) {
+			numRows = uint32(len(c.cs.pending.Rows))
+		}
+
+		// write first numRows
+		origRows := c.cs.pending.Rows
+		c.cs.pending.Rows = c.cs.pending.Rows[:numRows]
+		if err = c.writeBinaryRows(c.cs.pending); err != nil {
+			log.Errorf("Error writing result to %s: %v", c, err)
+			return err
+		}
+		if err = c.writeEndResult(false, 0, 0, handler.WarningCount(c)); err != nil {
+			log.Errorf("Error writing result to %s: %v", c, err)
+			return err
+		}
```

Comment:
Is this the right mechanics of the protocol? How does the client distinguish between: I fetched 128 rows, there were exactly 128 rows, and the cursor is now closed server-side. Versus: I fetched 128 rows, there are more than 128 rows available, the cursor is still open server-side?

Should writeEndResult() go at the end of every FETCH statement, or only when the result set is exhausted?